scanners/sslyze.py

```diff
@@ -180,62 +180,91 @@ def parse_sslyze(xml):
-        for group in groups:
-            size = int(group["GroupSize"])
-            if size < weakest_dh:
-                weakest_dh = size
+        for cipher in accepted_ciphers:
+            if cipher.get('dh_info', None) is not None:
+                size = int(cipher['dh_info']['GroupSize'])
+                if size < weakest_dh:
+                    weakest_dh = size
 
         if weakest_dh == 1234567890:
             weakest_dh = None
 
         data['config']['weakest_dh'] = weakest_dh
 
     # If there was an exception parsing the certificate, catch it before fetching cert info.
-    if target.select_one("certinfo[exception]") is not None:
-        data['errors'] = target.select_one("certinfo[exception]")["exception"]
+    if False:
+        data['errors'] = "TODO"
 
     else:
 
-        # Find the issuer of the last served cert.
-        # This is an attempt at finding the CA name, but won't work if the served
-        # chain is incomplete. I'll take what I can get without doing path chasing.
-        chain = target.select_one("certificateChain")
+        # Served chain.
+        served_chain = target['certinfo']['certificate_chain']
 
-        if not chain:
-            chain = target.select_one("receivedCertificateChain")
+        # Constructed chain may not be there if it didn't validate.
+        constructed_chain = target['certinfo']['verified_certificate_chain']
 
-        certificates = chain.select("certificate")
-        issuer = certificates[-1].select_one("issuer commonName")
-        if not issuer:
-            issuer = certificates[-1].select_one("issuer organizationalUnitName")
+        highest_served = parse_cert(served_chain[-1])
+        issuer = cert_issuer_name(highest_served)
 
-        if issuer and issuer.text:
-            data['certs']['served_issuer'] = issuer.text
+        if issuer:
+            data['certs']['served_issuer'] = issuer
         else:
             data['certs']['served_issuer'] = "(None found)"
 
-        leaf = chain.select_one("certificate[position=leaf]")
+        if (constructed_chain and (len(constructed_chain) > 0)):
+            highest_constructed = parse_cert(constructed_chain[-1])
+            issuer = cert_issuer_name(highest_constructed)
+            if issuer:
+                data['certs']['constructed_issuer'] = issuer
+            else:
+                data['certs']['constructed_issuer'] = "(None constructed)"
+
+        leaf = parse_cert(served_chain[0])
+        leaf_key = leaf.public_key()
 
-        # Key algorithm and length for leaf certificate only.
-        data['certs']['key_type'] = leaf.select_one("subjectPublicKeyInfo publicKeyAlgorithm").text
-        data['certs']['key_length'] = int(leaf.select_one("subjectPublicKeyInfo publicKeySize").text)
+        if hasattr(leaf_key, "key_size"):
+            data['certs']['key_length'] = leaf_key.key_size
+        elif hasattr(leaf_key, "curve"):
+            data['certs']['key_length'] = leaf_key.curve.key_size
```

Comment:
Filed https://github.com/pyca/cryptography/issues/3540 to track the issue that, if fixed, would remove the need for this ugly workaround.